dpc/src/record/encoded.rs

```diff
@@ -16,45 +16,93 @@
-    // TODO (howardwu): Remove this hardcoded value and use BaseField's size in bits to pad length.
-    let mut bytes = x_bytes.to_vec();
-    while bytes.len() < 32 {
-        bytes.push(0)
+    let mut res = Vec::<F>::with_capacity((bits.len() + capacity - 1) / capacity);
+    for chunk in bits.chunks(capacity) {
+        res.push(F::from_repr(F::BigInteger::from_bits_le(chunk)).unwrap());
     }
 
-    let input = P::BaseField::read_le(&bytes[..])?;
-    let (output, fq_high) = Elligator2::<P, G>::encode(&input)?;
-    Ok((output, fq_high))
+    Ok(res)
 }
 
-/// Decode a group into the byte representation of a base field element
-pub fn decode_from_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    affine: <G as ProjectiveCurve>::Affine,
-    fq_high: bool,
-) -> Result<Vec<u8>, DPCError> {
-    let output = Elligator2::<P, G>::decode(&affine, fq_high)?;
-    Ok(to_bytes_le![output]?)
+/// Decode a group into the byte representation of a base field element.
+pub fn decode_from_field<F: PrimeField>(field_elements: &[F]) -> Result<Vec<u8>, DPCError> {
+    if field_elements.is_empty() {
+        return Err(EncodingError::Message(
+            "The encoded record must consist of at least one field element.".to_string(),
+        )
+        .into());
+    }
+    // There is at least one field element due to the additional true bit.
+
+    let capacity = <F::Parameters as FieldParameters>::CAPACITY as usize;
+
+    let mut bits = Vec::<bool>::with_capacity(field_elements.len() * capacity);
+    for elem in field_elements.iter() {
+        let elem_bits = elem.to_repr().to_bits_le();
+        bits.extend_from_slice(&elem_bits[..capacity]); // only keep `capacity` bits, discarding the highest bit.
+    }
+
+    // Find the last bit of `true`, which should be within the last `CAPACITY` bits.
+    // If it does not exist, this is a `DPCError::EncodingError`.
+    let mut last_bit_pos = None;
+    for i in (0..bits.len()).rev().take(capacity) {
+        if bits[i] == true {
+            last_bit_pos = Some(i);
+            break;
+        }
+    }
+
+    if last_bit_pos.is_none() {
+        return Err(EncodingError::Message(
+            "The encoded record does not end with an expected termination bit.".to_string(),
+        )
+        .into());
+    }
+
+    bits.truncate(last_bit_pos.unwrap());
+    // This also removes the additional termination bit.
+
+    if bits.len() % 8 != 0 {
+        return Err(EncodingError::Message(
+            "The number of bits in the encoded record is not a multiply of 8.".to_string(),
+        )
+        .into());
+    }
+    // Here we do not use assertion since it can cause Rust panicking.
+
+    let mut bytes = Vec::with_capacity(bits.len() / 8);
+    for chunk in bits.chunks_exact(8) {
+        let mut byte = 0u8;
+        for bit in chunk.iter().rev() {
+            byte <<= 1;
+            byte += *bit as u8;
+        }
```

Comment:
I believe you can do something like this:
```rs
let bytes = u8::from_le_bytes(&chunk);
```

Reply:
That will also be a const function call then.

Reply:
nay, we need `from_bit_le` but it has not been implemented for `u8`.

Reply:
Ah shoot, good catch